### PR TITLE
Adding 3D Earth Model Capabilities

### LINKED ISSUE #32: Implement a 3D model of the Earth
Some Earth models are not spherically symmetric. We should have a 3D Earth class that is capable of generating neutrino paths in such models.

### PR DESCRIPTION
**Pull request by RPestes**
Resolves #32

### CODE REVIEW

File: EarthModelBase.cxx [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+/// @param lat - The latitude of the detector in deg N (between -90 and 90)
+/// @param lon - The longitude of the detector in deg E (between 0 and 360)
+///
+void EarthModelBase::SetDetPos(double rad, double lat, double lon)
+{
+
+  //Force radius to be non-negative
+  if(rad < 0)
+    rad = -rad;
+  fDetRadius = rad;
```
Comment (joaoabcoelho) [inline EarthModelBase.cxx:32]:
```suggestion
  if(rad < 0) {
    cerr << "WARNING: Negative radius detected. Setting to absolute value." << endl;
    rad = -rad;
  }
  fDetRadius = rad;
```

File: EarthModelBase.cxx [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+  //Force latitude to be between -90 and 90 deg
+  lat -= floor((lat+90.0)/360.0)*360.0;
+  if (lat > 90)
+    lat = 180.0-lat;
```
Comment (joaoabcoelho) [inline EarthModelBase.cxx:37]:
```suggestion
  if (lat > 90) {
    lat = 180.0-lat;
  }
```

File: EarthTables/model_defaults.hpp [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#include <string>
+const std::string MODEL_DIR = "/home/work/Desktop/AsymmetricEarth/EarthTables";
+const std::string PREM3D_DEFAULT = "/home/work/Desktop/AsymmetricEarth/EarthTables/prem_default.txt";
```
Comment (joaoabcoelho) [inline EarthTables/model_defaults.hpp:3]: You should not add paths like this, as it won't generalise to other users

Comment (RPestes) [reply EarthTables/model_defaults.hpp:3]: True... I didn't even think about that, since that's how it was in prem_defaults.hpp on my computer.  Is that a file that is configured when OscProb is installed?

File: PremModel.h [hunk 4]
```diff
@@ -27,42 +27,33 @@
...
       PremModel(std::string filename=""); ///< Constructor
       virtual ~PremModel();          ///< Destructor
 
-      virtual int FillPath(double cosT); ///< Fill the path sequence in a vector
+      int FillPath(double cosT, double phi=0); ///< Fill the path sequence in a vector
 
-      virtual std::vector<NuPath> GetNuPath(); ///< Get the current neutrino path sequence
-
-      virtual std::vector<NuPath> GetMergedPaths(double prec = 0.25); ///< Get merged path sequence in a vector
+//      virtual void SetDetPos(double pos); ///< Set the detector position in km
```
Comment (joaoabcoelho) [inline PremModel.h:47]:
```suggestion
```

File: EarthTables/model_defaults.hpp (hunk 3 above)
Comment (joaoabcoelho) [reply EarthTables/model_defaults.hpp:3]: Yes, this file is produced by the Makefile. It shouldn't even exist in the repository. See here:

https://github.com/joaoabcoelho/OscProb/blob/423ef0fbf0a15951f833f1b5cc48a82f5ae61aab/Makefile#L70-L73

File: EarthTables/prem_default.txt
Comment (joaoabcoelho) [inline EarthTables/prem_default.txt:1]: We should not change the default to 3D I think. Most users will not use the 3D version.

File: EarthTables/model_defaults.hpp (hunk 3 above)
Comment (RPestes) [reply EarthTables/model_defaults.hpp:3]: OK.  I will remove it, then, modify EarthModelBinned.cxx to use prem_default.hpp, and add lines to the makefile to have it add variables for the EarthTables directory and the 3D version of prem_default.txt to prem_default.hpp.  Does that sound good to you?

File: EarthTables/prem_default.txt
Comment (RPestes) [reply EarthTables/prem_default.txt:1]: This would just be the default table to use within the EarthModelBinned class.

File: EarthTables/model_defaults.hpp (hunk 3 above)
Comment (RPestes) [reply EarthTables/model_defaults.hpp]: I finished making these changes...

File: Makefile [hunk 5]
```diff
@@ -16,7 +16,7 @@ else
   DICTFLAGS = -c -p
 endif
 
-PACKAGE = OscProb
+PACKAGE = OscProb2
```
Comment (joaoabcoelho) [inline Makefile:19]: Why this change?

File: Makefile [hunk 6]
```diff
@@ -70,7 +72,9 @@ $(CLEANDIRS):
 $(PREMINC): $(PREMDIR) $(PREMFILE)
```
Comment (joaoabcoelho) [inline Makefile:72]: You should add `$(MODEL3DDIR)` and `$(PREM3DFILE)` to the dependencies

File: Makefile (hunk 5 above)
Comment (RPestes) [reply Makefile:19]: Oops... I wasn't trying to make that change.  I'll change it back.  It was that way on my computer because I was trying to keep this modified package separate from the original OscProb package.

File: EarthModelBinned.cxx [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+
+  fEarthBins.push_back( EarthBin(radius_out, radius_in, latitude, longitude, density, zoa, index) );
+
+}
+
+//......................................................................
+///
+/// Load an earth model from a file.
+///
+/// By default it loads the model stored in EarthTables/prem_default.txt
```
Comment (joaoabcoelho) [inline EarthModelBinned.cxx:89]:
```suggestion
/// By default it loads the model stored in EarthTables/earth_binned_default.txt
```

File: EarthModelBinned.cxx [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+
+using namespace std;
+
+using namespace OscProb;
+
+//......................................................................
+///
+/// Constructor.
+///
+/// By default this implements the model stored in EarthTables/prem_default.txt
```
Comment (joaoabcoelho) [inline EarthModelBinned.cxx:22]:
```suggestion
/// By default this implements the model stored in EarthTables/earth_binned_default.txt
```

File: EarthModelBinned.h [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+///
+/// The detector can be positioned at any point within the Earth, and
+/// the path sequences will take into account the fact that some layers
+/// are above the detector.
+///
+/// Using indices to specify various regions (which must have a constant
+/// Z/A value), the Z/A value for a region can be changed, and the density
+/// throughout the region can be scaled by a constant factor.
+///
+/// By default this implements the model stored in EarthTables/prem_default.txt
```
Comment (joaoabcoelho) [inline EarthModelBinned.h:21]:
```suggestion
/// By default this implements the model stored in EarthTables/earth_binned_default.txt
```

File: EarthModelBinned.h [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+      virtual double DetDistForNextLonBin(double prev_lon, double dLon, int &lon_bin, double min_lon, double max_lon, double sinDetLon, double cosDetLon, double alpha, double sinTsinAsinDetLon, double sinTsinAcosDetLon, double rDetCosDetLat); ///< Calculate the detector distance at the edge of the current lon bin along the neutrino's trajectory and increment lon_bin
+
+      virtual void RecordLatLonBinCrossings(double detDist_nextDbin, double &DetDist, int &index, int &latBin, int &nextLatBin, double &detDist_nextLatBin, int &sign, double dLat, double maxSinSqLat, int &maxlatreached, int &lonBin, int &nextLonBin, double &detDist_nextLonBin, double dLon, double min_lon, double max_lon, double alpha, double beta, double gamma, double gammaSq, double rDetGammaSinDetLat, double rDetCosAcosDetLat, double rDetSinDetLat, double rDetCosDetLat, double rDetCosT, double rDetSinT, double sinTsinAsinDetLon, double sinTsinAcosDetLon, double sinDetLon, double cosDetLon); ///< Record path segments for each latitude/longitude bin crossed before reaching detDist_nextDbin
```
Comment (joaoabcoelho) [inline EarthModelBinned.h:148]: Many of these functions have too many arguments. Especially in C++, this is not great because you need them to be input in the exact order they were declared. Instead you should create some classes or structs to represent the data you want to pass in or out.

Comment (RPestes) [reply EarthModelBinned.h]: Many of the arguments for DetDistForNextLatBin, DetDistForNextLonBin, and RecordLatLonBinCrossings have now been replaced by arguments of structs LatBinInfo and LonBinInfo.  I also created a struct (TrajConstants) to hold most of the variables that are different combinations of the detector coordinates and neutrino trajectory angles.  An private object of that struct is included in the EarthModelBinned class, so those variables are no longer included as arguments of functions.  Does that address all your concerns about this?

Comment (joaoabcoelho) [reply EarthModelBinned.h]: Looks, much better. If all tests pass I will merge this pull request. Thanks a lot!
